### PR TITLE
550 - Implement SASL support for LDAP

### LINKED ISSUE #550: [Support] Using Keytab files (SASL)  for authentication
**Environment:**
 - LDAP Server Type: ActiveDirectory / OpenLDAP / FreeIPA
 - PHP Version: 7.3 / 7.4 / 8.0 / 8.1 /8.2

At Passbolt we want to be able to use a Keytab file to authenticate against the LDAP server. So I have defined the following steps to achieve it:

* Extend LdapRecord/Ldap
* Overwrite bind to use ldap_sasl_bind
* (May need to) Extend Guard class to pass correct parameters to `connection->bind`
* In our `getConnection()` method we need to pass new Ldap object as second parameter to `Connection` object if authentication is set to use Keytab. Otherwise keep current behavior.

So my questions are:
* Do you think I am missing something on this approach? or you see an easier way to achieve the same?
* Do you find this feature useful? Is it something you would like us to contribute to the plugin?

Thank you!

### PR DESCRIPTION
**Pull request by ajibarra**
Fixes #550 . Please let me know if the approach is correct or if anything else is needed.

### CODE REVIEW

**Comment by ajibarra** [comment]
@stevebauman I have updated the unit tests to fix issue reported in previous run

File: src/Ldap.php [hunk 1]
```diff
@@ -300,10 +300,23 @@ public function parseResult($result, &$errorCode, &$dn, &$errorMessage, &$referr
...
+                'authz_id' => null,
+                'props' => null,
+            ];
+            $saslOptions = array_merge($defaultSaslOptions, $saslOptions);
+        }
+
+        return $this->bound = $this->executeFailableOperation(function () use ($username, $password, $sasl, $saslOptions) {
+            return $sasl ?
+                ldap_sasl_bind($this->connection, $username, $password, $saslOptions['mech'], $saslOptions['realm'], $saslOptions['authc_id'], $saslOptions['authz_id'], $saslOptions['props'])
+                : ldap_bind($this->connection, $username, $password ? html_entity_decode($password) : null);
```
Comment (stevebauman) [inline src/Ldap.php:319]: We should separate this into its own method (maybe `saslBind()`, as the new major version replaces `ldap_bind()` `ldap_bind_ext()`, which returns a different type (`LDAPResult` vs `bool`):

https://github.com/DirectoryTree/LdapRecord/blob/a4c95f3a16b73d59c21a10c993562ff9baeee661/src/Ldap.php#L287-L298

Thoughts?

Comment (ajibarra) [reply src/Ldap.php:319]: @stevebauman 

Yes that would work so we would need:
1. Change Guard to call the required method (bind or saslBind)
2. Add saslBind to LdapInterface and implement it on Ldap and LdapFake

Let me know if this is the approach you were thinking of. Thank you

Comment (stevebauman) [reply src/Ldap.php:319]: @ajibarra Yup that totally works for me! 👍

**Pull request closed without merging**